MDEV-29091: Correct event_name in PFS for wait caused by FOR UPDATE

haidong opened the pull request:
- [x] *The Jira issue number for this PR is: MDEV-29091*
## Description
When one session SELECT rows FOR UPDATE and holds the lock, subsequent
sessions that SELECT ... FOR UPDATE will wait to get the lock.
Currently, that event is labeled as `wait/io/table/sql/handler`, which
is incorrect. Because it is NOT waiting for IO. Instead, it should have
been `wait/lock/table/sql/handler`.

## How can this PR be tested?
All existing test cases should still be passing in Buildbot as well as
GitLab-CI. In addition, new test case added for this PR should also pass.

## Basing the PR against the correct MariaDB version
- [x] This is a big fix and the PR is based against the long-term support release.

## Backward compatibility
This bug fix should be fully backward-compatible

## Copyright
All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.

grooverdan commented:
Nice find and fix.

Does the test [events_waits_current_MDEV-29091.test](https://github.com/MariaDB/server/pull/2206/files#diff-9f24e53bc956104818d1d63ff73b73535da2f3811ec58a96b70d038a23933fcd) test something not covered in the changed results of the other tests?
If no, can it be removed?
If yes;
* please replace the `sleep()` statements with `let $wait_condition=... --source include/wait_condition.inc`
* remove "IF EXISTS" from table setup/cleanup. A clean environment can be assumed.
* do other perf tests cleanup `setup_instruments`/`setup_consumers`?

Thanks for the great bug report/commit message.

haidong commented:
Thanks @grooverdan. Great comments and suggestions. I have addressed them all, see below. The changes have also been pushed in.

>> Does the test [events_waits_current_MDEV-29091.test](https://github.com/MariaDB/server/pull/2206/files#diff-9f24e53bc956104818d1d63ff73b73535da2f3811ec58a96b70d038a23933fcd) test something not covered in the changed results of the other tests?

Yes, it covers the specific scenario of `FOR UPDATE` lock and then wait. This is not covered by other tests under the perfschema suite or elsewhere.

- `sleep()` is removed. Thanks!
- Ditto
- It is not consistently cleaned up, most don't. I have made changes to `setup_instruments/setup_consumers` more targeted, and added the cleanup step at the end.

grooverdan commented:
related; http://datavirtualizer.com/are-some-mysql-wait-events-actually-cpu-time/

haidong commented:
> related; http://datavirtualizer.com/are-some-mysql-wait-events-actually-cpu-time/

Good find. Yes, like the article says, performance_schema currently labels way too many threads waiting on `wait/io/table/sql/handler`. I think this PR should help mitigate that problem quite a bit.

Like I mentioned in the commit message, when we have time, I think some rethinking and refactoring of `TABLE_IO_WAIT` and its downstream functions will be useful.

haidong commented:
I have [asked for Vladislav Lesin's input](https://jira.mariadb.org/browse/MDEV-29091) a month ago on the JIRA page, but haven't received a reply.

@LinuxJedi Could you take a look, and if not objections, merge it? I believe this will mitigate the exaggeration of `wait/io/tables/sql/handler` considerably.

LinuxJedi commented:
@haidong personally I have no objections to it, but it should probably be against 10.3 since that is the oldest LTS. I'll ping @vlad-lesin about this internally.

haidong commented:
I've applied the code to 10.3 branch. The PR has also been adjusted so it's against 10.3, not 10.6.

The [`amd64-ubuntu-2004-debug` failure](https://buildbot.mariadb.org/#builders/369/builds/5154) baffles me. The test failed is ` rpl.rpl_parallel`. This test:
- should not have anything to do with this change;
- was successful in a prior CI run;
- didn't fail in other builds in the same CI run;
- was also successful on my dev desktop when I ran it locally.

@LinuxJedi or @grooverdan, any suggestions?

LinuxJedi commented:
Rebased off 10.4 as 10.3 is near EoL and this commit would not land there.

vuvova commented:
Could, please, anyone explain what this patch is doing? What does `wait->m_class` mean before it's initialized?

LinuxJedi commented:
After discussing this with Sergei and my own testing, we have decided to revert this.

Whilst in this specific scenario it does show a table lock, it appears to be more of a happy accident than it really showing the correct lock. The `FOR UPDATE` itself does not appear to trigger the correct path for this, and it is likely more instrumentation to PFS will need to be added to get the desired outcome.
